xds/src/main/java/io/grpc/xds/FilterChainMatchingHandler.java

```diff
@@ -0,0 +1,362 @@
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Strings;
+import com.google.common.collect.Iterables;
+import com.google.protobuf.UInt32Value;
+import io.grpc.Attributes;
+import io.grpc.internal.ObjectPool;
+import io.grpc.netty.GrpcHttp2ConnectionHandler;
+import io.grpc.netty.InternalProtocolNegotiationEvent;
+import io.grpc.netty.InternalProtocolNegotiator;
+import io.grpc.netty.InternalProtocolNegotiator.ProtocolNegotiator;
+import io.grpc.netty.ProtocolNegotiationEvent;
+import io.grpc.xds.EnvoyServerProtoData.CidrRange;
+import io.grpc.xds.EnvoyServerProtoData.ConnectionSourceType;
+import io.grpc.xds.EnvoyServerProtoData.FilterChain;
+import io.grpc.xds.EnvoyServerProtoData.FilterChainMatch;
+import io.grpc.xds.internal.Matchers.CidrMatcher;
+import io.grpc.xds.internal.sds.SslContextProviderSupplier;
+import io.netty.channel.ChannelHandler;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelInboundHandlerAdapter;
+import io.netty.util.AsciiString;
+import java.net.Inet6Address;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.Executor;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.annotation.Nullable;
+
+
+/**
+ * Handles L4 filter chain match for the connection based on the xds configuration.
+ * */
+final class FilterChainMatchingHandler extends ChannelInboundHandlerAdapter {
+  private static final Logger log = Logger.getLogger(FilterChainMatchingHandler.class.getName());
+
+  private static final AsciiString SCHEME = AsciiString.of("http");
+
+  private final GrpcHttp2ConnectionHandler grpcHandler;
+  private final FilterChainSelector selector;
+  private final ProtocolNegotiator delegate;
+
+  FilterChainMatchingHandler(
+          GrpcHttp2ConnectionHandler grpcHandler, FilterChainSelector selector,
+          ProtocolNegotiator delegate) {
+    this.grpcHandler = checkNotNull(grpcHandler, "grpcHandler");
+    this.selector = checkNotNull(selector, "selector");
+    this.delegate = checkNotNull(delegate, "delegate");
+  }
+
+  @Override
+  public void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exception {
+    if (!(evt instanceof ProtocolNegotiationEvent)) {
+      super.userEventTriggered(ctx, evt);
+      return;
+    }
+    SelectedConfig config = selector.select(
+            (InetSocketAddress) ctx.channel().localAddress(),
+            (InetSocketAddress) ctx.channel().remoteAddress());
+    if (config == null) {
+      log.log(Level.FINER, "Did not find exactly one filter chain.");
+      ctx.fireExceptionCaught(
+              new IllegalStateException("Did not find exactly one filter chain."));
```

Comment:
With `FINER` logging level one would know  but I would highly recommend stating the exact failure reason instead of "Did not find exactly one ...". Was it 0 or was it 3 matches? If we are unable to communicate that thru return value can we use a thrown exception around `selector.select`?

Reply:
Im not entirely sure. It seems easier for user to handle exception if we just do `ctx.fireExceptionCaught`, but we also prefer error messages in exception to be more specific.
Maybe we could throw `selector.select` then catch here(similarly we can return `ErrorOrStructure`), just it would be a bit cumbersome than the current null + log. But anyways I hope this is not a blocker.

Reply:
May be I misspoke: I *was* thinking about the message in the `IllegalStateException` to reflect the exact case instead of the more ambiguous "not exactly one". It will be highly desirable if you can do it